clients/drcachesim/tools/filter/record_filter.cpp

```diff
@@ -521,6 +634,36 @@ record_filter_t::parallel_shard_memref(void *shard_data, const trace_entry_t &in
     ++per_shard->input_entry_count;
     trace_entry_t entry = input_entry;
     bool output = true;
+    // XXX: Once we have multi-workload inputs we'll want all our PC keys to become
+    // pairs <get_workload_ordinal(), PC>.
+    if (per_shard->shard_stream->get_workload_id() != per_shard->prev_workload_id &&
+        per_shard->shard_stream->get_workload_id() >= 0 &&
+        per_shard->prev_workload_id >= 0) {
+        per_shard->error = "Multi-workload inputs not yet supported";
+        return false;
+    }
+    int64_t input_id = per_shard->shard_stream->get_input_id();
+    if (per_shard->prev_input_id != input_id) {
+        VPRINT(this, 3,
+               "shard %d switch from %" PRId64 " to %" PRId64 " (refs=%" PRIu64
+               " instrs=%" PRIu64 ")\n",
+               per_shard->shard_stream->get_shard_index(), per_shard->prev_input_id,
+               input_id,
+               per_shard->shard_stream->get_input_interface() == nullptr
+                   ? 0
+                   : per_shard->shard_stream->get_input_interface()->get_record_ordinal(),
+               per_shard->shard_stream->get_input_interface() == nullptr
+                   ? 0
+                   : per_shard->shard_stream->get_input_interface()
+                         ->get_instruction_ordinal());
+        std::lock_guard<std::mutex> guard(input2info_mutex_);
+        auto it = input2info_.find(input_id);
+        if (it == input2info_.end()) {
+            input2info_[input_id] = std::unique_ptr<per_input_t>(new per_input_t);
+            it = input2info_.find(input_id);
+        }
+        per_shard->per_input = it->second.get();
```

Comment:
Do we need to add a debug check that nullifies it->second in any other shard data structure? Maybe that'll be too expensive.

Reply:
Added a comment.